tensorflow/core/kernels/stack.cc

```diff
@@ -72,15 +78,85 @@ class Stack : public ResourceBase {
       return errors::InvalidArgument("Stack[", stack_name_,
                                      "] is empty when calling Pop().");
     }
+    int index = stack_.size() - 1;
+    while (is_swapping_ins_[index]) {
+      cond_swapping_ins_[index].wait(l);
+      index = stack_.size() - 1;
+    }
     *value = stack_.back();
     stack_.pop_back();
     return Status::OK();
   }
 
+  Status SwapOut(TensorAndAllocation* value, Tensor* cpu_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *cpu_tensor;
+    value->swapped_to_cpu = true;
+
+    return Status::OK();
+  }
+
+  Status SwapIn(TensorAndAllocation* value, Tensor* device_tensor) {
+    mutex_lock l(mu_);
+    TF_RETURN_IF_ERROR(CheckNotClosed());
+
+    value->tensor = *device_tensor;
+    value->swapped_to_cpu = false;
+
+    return Status::OK();
+  }
+
+  bool GetTensorToSwapOut(
+      TensorAndAllocation** value,
+      std::function<bool(const TensorAndAllocation&)> cond) {
+    mutex_lock l(mu_);
+    bool can_swap = false;
+    while (!can_swap) {
+      if (unswapped_ids_.empty()) {
+        break;
+      }
+      int index = unswapped_ids_.front();
+      unswapped_ids_.pop();
+      if (!cond(stack_[index])) {
+        continue;
+      }
+      swapped_ids_.push(index);
+      *value = &stack_[index];
+      can_swap = true;
+    }
+    return can_swap;
+  }
+
+  bool GetTensorToSwapIn(TensorAndAllocation** value, int& index) {
```

Comment:
Google C++ style guide says the argument should be int* index since it's being modified:
https://google.github.io/styleguide/cppguide.html#Output_Parameters

Reply:
Thanks for letting me know, noted.